File: src/main/scala/com/vc/hard/L52.scala
package com.vc.hard

object L52 {
  def totalNQueens(n: Int): Int = {
    var res = 0
    val arr = Array.ofDim[Char](n, n)
    arr.indices.foreach(i => {
      arr(i).indices.foreach(j => {
        arr(i)(j) = '.'
      })
    })

    def solve(row: Int): Unit = {
      if(row == n) res += 1
      else {
        (0 until n).foreach(col => {
          arr(row)(col) = 'Q'
          if(isValid(row, col)) solve(row + 1)
          arr(row)(col) = '.'
        })
      }
    }

    def isValid(row: Int, col: Int): Boolean = {
      var i = 0
      var j = col
      while(i < row) {
        if(arr(i)(j) == 'Q') return false
        i += 1
      }

      i = row - 1
      j = col - 1
      while(i >= 0 && j >= 0) {
        if(arr(i)(j) == 'Q') return false
        i -= 1
        j -= 1
      }

      i = row - 1
      j = col + 1
      while(i >= 0 && j < n) {
        if(arr(i)(j) == 'Q')  return false
        i -= 1
        j += 1
      }
      true
    }
    solve(0)
    res
  }
}

File: src/main/java/com/vc/hard/L52.java
package com.vc.hard;

import java.util.HashSet;

class L52 {
    int res = 0;
    HashSet<Integer> cols = new HashSet<>();
    HashSet<Integer> diagonal = new HashSet<>();
    HashSet<Integer> antidigonal = new HashSet<>();

    public int totalNQueens(int n) {
        char[][] state = new char[n][n];
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                state[i][j] = '.';
            }
        }
        helper(0, n, state);
        return res;
    }

    private void helper(int row, int n, char[][] state) {
        if(row == n) res += 1;

        for(int col = 0; col < n; col++) {
            if(isValid(row, col, n, state)) {
                state[row][col] = 'Q';
                cols.add(col);
                diagonal.add(row + col);
                antidigonal.add(col - row);

                helper(row + 1, n, state);

                state[row][col] = '.';
                cols.remove(col);
                diagonal.remove(row + col);
                antidigonal.remove(col - row);
            }
        }
    }

    private boolean isValid(int row, int col, int n, char[][] state) {
        if(cols.contains(col)) return false;
        if(diagonal.contains(row + col)) return false;
        if(antidigonal.contains(col - row)) return false;
        return true;
    }
}